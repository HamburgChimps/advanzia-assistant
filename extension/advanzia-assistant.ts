export interface ContentScript {
  execute(): void;
}

export enum ContentScriptStatus {
  Initializing = 'Initializing',
  Noop = 'Noop',
  Ready = 'Ready',
  Done = 'Done',
  Err = 'Error',
}

export interface WasmExports extends WebAssembly.Exports {
  add_one: (num: number) => number,
}

export interface ContentScriptEvents {
  readonly noop: Event;
  readonly ready: Event;
  readonly done: Event;
  readonly error: ErrorEvent;
}

export class Script extends EventTarget implements ContentScript, EventListenerObject {
  private memory: WebAssembly.Memory;

  private wasmExports?: WasmExports;

  private events: ContentScriptEvents;

  status: ContentScriptStatus;

  errors: Error[];

  constructor() {
    super();
    this.status = ContentScriptStatus.Initializing;
    this.memory = new WebAssembly.Memory({ initial: 10 });
    this.events = {
      noop: new Event('noop'),
      ready: new Event('ready'),
      done: new Event('done'),
      error: new ErrorEvent('error'),
    };

    this.errors = [];

    this.registerEventListeners();
  }

  private registerEventListeners() {
    Object.values(this.events).forEach((e) => this.registerEventListener(e.type));
  }

  private registerEventListener(name: string) {
    this.addEventListener(name, this);
  }

  handleEvent(e: Event | ErrorEvent) {
    switch (e.type) {
      case this.events.ready.type:
        this.status = ContentScriptStatus.Ready;
        break;
      case this.events.noop.type:
        this.status = ContentScriptStatus.Noop;
        break;
      case this.events.done.type:
        this.status = ContentScriptStatus.Done;
        break;
      case this.events.error.type:
        this.status = ContentScriptStatus.Err;
        break;
      default:
        break;
    }
  }

  on(eventName: string, handler: EventListenerOrEventListenerObject) {
    this.addEventListener(eventName, handler);
    return this;
  }

  async execute() {
    if (window.location.host !== 'mein.gebuhrenfrei.com') {
      this.error(new Error('this script can only be run on mein.gebuhrenfrei.com'));
      return;
    }

    if (window.location.pathname.indexOf('retail-app') === -1) {
      this.dispatchEvent(this.events.noop);
      return;
    }

    const wasmPath = chrome.runtime.getURL('advanzia-assistant.wasm');
    const wasmResponse = await fetch(wasmPath);

    const wasmResponseArrayBuffer = await wasmResponse.arrayBuffer();

    const wasmImports = { env: { memory: this.memory } };

    const wasm = await WebAssembly.instantiate(wasmResponseArrayBuffer, wasmImports);
    this.wasmExports = wasm.instance.exports as WasmExports;

    await Script.pageReasonablyLoaded();

    this.dispatchEvent(this.events.ready);
  }

  private error(e: Error) {
    this.errors.push(e);
    this.dispatchEvent(this.events.error);
  }

  private static pageReasonablyLoaded(): Promise<void> {
    const signalSelector = '.card';
    return new Promise((resolve) => {
      const el = document.querySelector(signalSelector);
      if (el) {
        resolve();
        return;
      }
      new MutationObserver((_mutationsList, observer) => {
        if (document.querySelector(signalSelector)) {
          resolve();
          observer.disconnect();
        }
      })
        .observe(document.documentElement, {
          childList: true,
          subtree: true,
        });
    });
  }
}
