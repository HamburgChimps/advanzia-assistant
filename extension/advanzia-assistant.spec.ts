import * as fs from 'fs';
import * as chromeMock from 'jest-chrome';
import { Response } from 'cross-fetch';
import { ContentScript, Script } from './advanzia-assistant';

describe('Content Script', () => {
    const eventsRaised: { [key: string]: boolean } = {
        noop: false,
        ready: false,
        done: false,
        error: false,
    };

    const setRaised = (eventName: string) => () => { eventsRaised[eventName] = true };
    const resetRaised = (eventName: string) => { eventsRaised[eventName] = false };
    const resetAllRaised = () => Object.keys(eventsRaised).forEach(resetRaised);

    let script: ContentScript;

    beforeAll(() => Object.assign(global, chromeMock));
    beforeEach(resetAllRaised);
    beforeEach(() => {
        script = new Script()
            .on('noop', setRaised('noop'))
            .on('ready', setRaised('ready'))
            .on('done', setRaised('done'))
    });
    beforeEach(jest.resetAllMocks);
    beforeEach(jest.restoreAllMocks);
    describe('when not on transactions page', () => {
        it('should not do anything', async () => {
            await script.execute();

            expect(chrome.runtime.getURL).toBeCalledTimes(0);
            expect(eventsRaised.noop).toBe(true);
            expect(eventsRaised.ready).toBe(false);
            expect(eventsRaised.done).toBe(false);
            expect(eventsRaised.error).toBe(false);
        });
    });

    describe('when on transactions page', () => {
        const simulateAdvanziaAppDoingStuff = () => setTimeout(() => {
            const div = document.createElement('div');
            div.className = 'card';
            document.body.appendChild(div);
        }, 2000);

        beforeEach(() => {
            jest.spyOn(window, "location", "get").mockReturnValue({
                ...window.location,
                ...{ pathname: '/retail-app-de' },
            });
            global.fetch = async () => new Response(fs.readFileSync('../dist/advanzia-assistant.wasm'));
        });

        beforeEach(simulateAdvanziaAppDoingStuff);
        beforeEach(() => script.execute());
        it('should indicate ready status', async () => {
            expect(chrome.runtime.getURL).toBeCalledTimes(1);
            expect(eventsRaised.noop).toBe(false);
            expect(eventsRaised.ready).toBe(true);
            expect(eventsRaised.done).toBe(false);
            expect(eventsRaised.error).toBe(false);
        });

        it('to insert correct dom modifications', async () => {
            expect(document.body).toMatchSnapshot();
        });
    })
});