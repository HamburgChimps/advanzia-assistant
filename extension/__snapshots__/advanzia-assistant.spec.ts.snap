// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Script when on transactions page to insert correct dom modifications 1`] = `
<body>
  <div
    class="card"
  />
</body>
`;
